src/corona_hakab_model/affinity_matrix.py

```diff
@@ -212,17 +88,169 @@ def normalize(self):
+                remaining_contacts[second_agent] -= 1
+                connections_sum -= 2
+
+                if remaining_contacts[second_agent] <= 0:
+                    del (available_agents[roll_index])
+                    available_agents_len -= 1
+                else:
+                    temp_agents_holder.append(available_agents.pop(roll_index))
+                    available_agents_len -= 1
+            # returning all the agents back from the temp place holder
+            available_agents_len += len(temp_agents_holder)
+            available_agents.extend(temp_agents_holder)
+
+        # filling the matrix. sometimes there still remains 1 un-filled connection, and it is left as 0,0 connection, so reset 0,0
+        matrix[connections_tuples[0], connections_tuples[1]] = connection_strength
+        matrix[0, 0] = 0
+
+        return matrix
+
+    def circular_matrix_generation(self, agents, circle_size_probability: rv_discrete, connection_strength):
+        """
+        this method will create a matrix of circular connections given a list of agents, an rv_discrete representing the size of the circle probabilty, and the connection strength
+        :param agents: list of all agents
+        :param circle_size_probability: representing the size probability
+        :param connection_strength: the wanted connection strength
+        :return: lil_matrix with the wanted connections
+        """
+        my_agents_length = len(agents)
+        matrix = m_type((self.size, self.size), dtype=np.float32)
+
+        # pre-rolling all the sized of the circles for efficiency causes.
+        circles_size_rolls = circle_size_probability.rvs(
+            size=math.ceil(my_agents_length / circle_size_probability.mean()))
+        pre_rolled_circles_len = len(circles_size_rolls)
+
+        circles_counter = 0
+        agents_counter = 0
+
+        # here all the circles will be saved
+        circles: List[TrackingCircle] = []
+
+        agents_ids = np.array([a.index for a in agents])
+        np.random.shuffle(agents_ids)
+
+        # loop creating all circles. each run creates one circle.
+        while agents_counter < my_agents_length:
+
+            current_circle = TrackingCircle()
+
+            # choosing current circle size
+            current_circle_size = 0
+            if circles_counter < pre_rolled_circles_len:
+                current_circle_size = circles_size_rolls[circles_counter]
+                circles_counter += 1
+            else:
+                current_circle_size = circle_size_probability.rvs()
+
+            # adding agents to the current circle
+            for _ in range(current_circle_size):
+                if agents_counter >= my_agents_length:
+                    break
+                choosen_agent = agents[agents_ids[agents_counter]]
+                current_circle.add_agent(choosen_agent)
+                # todo possibly add this circle to the agent circles list
+                agents_counter += 1
```

Comment:
why not pop from agent_ids and spare the trouble?

Reply:
len simply retrieves a stored value from the list

Reply:
ok, switched to len